src/ts/features/block-navigation-mode/index.ts

```diff
@@ -0,0 +1,410 @@
+import {Feature, Settings} from '../../utils/settings'
+import {clearBlockNavigationView} from './blockNavigationView'
+import {
+    ensureFocusedPanelHasBlockSelected,
+    firstNativelyHighlightedBlock,
+    jumpBlocksInFocusedPanel,
+    jumpUntilSelectedBlockIsVisible,
+    lastNativelyHighlightedBlock,
+    scrollFocusedPanel,
+    scrollUntilBlockIsVisible,
+    selectedBlock,
+    selectedBlockId,
+    state,
+} from './blockNavigation'
+import {Selectors} from '../../roam/roam-selectors'
+import {Mouse} from '../../utils/mouse'
+import {initializeBlockNavigationMode} from './blockNavigationInit'
+import {map, Mode, nimap, nmap, nvmap, returnToNormalMode} from './vim'
+import {getHint, HINT_IDS, HINT_KEYS} from './blockNavigationHintView'
+import {Roam} from '../../roam/roam'
+import {Keyboard} from '../../utils/keyboard'
+import {KEY_TO_SHIFTED} from '../../utils/react-hotkeys'
+import {copyBlockEmbed, copyBlockReference} from '../../roam/roam-block'
+import {delay} from '../../utils/async'
+
+const _jumpBlocksInFocusedPanel = async (mode: Mode, blocksToJump: number) => {
+    if (mode == 'NORMAL') {
+        jumpBlocksInFocusedPanel(blocksToJump)
+        scrollUntilBlockIsVisible()
+    }
+    if (mode == 'VISUAL') {
+        for (let i = 0; i < Math.abs(blocksToJump); i++) {
+            // TODO figure out why simulating 50 key presses doesn't really work for visual mode
+            // there's probably a better way
+            await Keyboard.simulateKey(blocksToJump > 0 ? Keyboard.DOWN_ARROW : Keyboard.UP_ARROW, 0, {shiftKey: true})
+        }
+        scrollUntilBlockIsVisible(blocksToJump > 0 ? lastNativelyHighlightedBlock() : firstNativelyHighlightedBlock())
+    }
+}
+
+const insertBlockAfter = async () => {
+    const block = selectedBlock()
+    if (block) {
+        await Roam.activateBlock(block)
+        await Roam.createBlockBelow()
+    }
+}
+
+const cutAndGoBackToNormal = async () => {
+    document.execCommand('cut')
+    // Wait for the block to disappear, double check that a block is still selected
+    // Deleting the first line can lead to no previous block existing to select
+    await delay(0)
+    ensureFocusedPanelHasBlockSelected()
+    await returnToNormalMode()
+}
+
+export const config: Feature = {
```

Comment:
majority of the things here should probably be separate functions in respective modules vs being done inline